feat: Add custom completer for `_cargo_example_names`

shannmu opened the pull request:
<!-- homu-ignore:start -->
<!--
Thanks for submitting a pull request 🎉! Here are some tips for you:

* If this is your first contribution, read "Cargo Contribution Guide" first:
  https://doc.crates.io/contrib/
* Run `cargo fmt --all` to format your code changes.
* Small commits and pull requests are always preferable and easy to review.
* If your idea is large and needs feedback from the community, read how:
  https://doc.crates.io/contrib/process/#working-on-large-features
* Cargo takes care of compatibility. Read our design principles:
  https://doc.crates.io/contrib/design.html
* When changing help text of cargo commands, follow the steps to generate docs:
  https://github.com/rust-lang/cargo/tree/master/src/doc#building-the-man-pages
* If your PR is not finished, set it as "draft" PR or add "WIP" in its title.
* It's ok to use the CI resources to test your PR, but please don't abuse them.

### What does this PR try to resolve?

Explain the motivation behind this change.
A clear overview along with an in-depth explanation are helpful.

You can use `Fixes #<issue number>` to associate this PR to an existing issue.

### How should we test and review this PR?

Demonstrate how you test this change and guide reviewers through your PR.
With a smooth review process, a pull request usually gets reviewed quicker.

If you don't know how to write and run your tests, please read the guide:
https://doc.crates.io/contrib/tests

### Additional information

Other information you want to mention in this PR, such as prior arts,
future extensions, an unresolved problem, or a TODO list.
-->
<!-- homu-ignore:end -->
Replace `_cargo_example_names` to the custom completer

Based on https://github.com/shannmu/cargo/pull/2

0xPoe committed:
feat: Add `info` cargo subcommand

ifropc committed:
feat: add new lockfile-path for metadata

- Also add necessary tests for the new flag

0xPoe committed:
test: migrate all test from cargo-information

ifropc committed:
feat: add lockfile-path flag for all requested commands

- All commands that support `manifest-path` (except `locate-project`, `verify-project` and `read-manifest`)

0xPoe committed:
docs: add cargo-info subcommand and add man page

0xPoe committed:
docs: add completions for bash and zsh

linyihai committed:
feat: Add matches_prerelease semantic

linyihai committed:
doc: Add mod level documents

ifropc committed:
test: add extra test for package command

linyihai committed:
feat: add upper bound semantic for `Less Op`

weihanglo committed:
refactor: extract remap rules into functions

ifropc committed:
doc: add docs for lockfile-path

ifropc committed:
test: lockfile path test refactoring

- win/mac fix

ifropc committed:
refactoring: lockfile-path documentation and implementation cleanup

ifropc committed:
feat: lockfile path refactoring

ifropc committed:
refactoring: move lock_root into worspace

dpaoliello committed:
Implement base paths (RFC 3529) 1/n: path dep and patch support

ifropc committed:
feat lockfile-path and related tests refactoring

weihanglo committed:
feat(trim-paths): rustdoc supports trim-paths for diagnostics

This enables rustdoc invocation from `cargo doc` to trim paths
in diagnostics. Doc tests are out-of-scope, as they are handled
in a different mechanism, which needs more cares.

jemtucker committed:
test: Add test for issue rust-lang#14379

jemtucker committed:
fix: use longhand gitoxide path-spec patterns

bors committed:
Auto merge of #14141 - hi-rustin:rustin-patch-info, r=epage

feat: Add `info` cargo subcommand

<!-- homu-ignore:start -->

close https://github.com/rust-lang/cargo/issues/14081
close https://github.com/rust-lang/cargo/issues/948

fcp https://github.com/rust-lang/cargo/pull/14141#issuecomment-2243683924

This PR added a new `info` cargo subcommand.

# Background

This adds a new subcommand to Cargo, `cargo info`. This subcommand would allow users to get information about a crate from the command line, without having to go to the web.

The main motivation for this is to make it easier to get information about a crate from the command line. Currently, the way to get information about a crate is to go to the web and look it up on [crates.io] or find the crate's source code and look at the `Cargo.toml` file. This is not very convenient, especially not all information is displayed on the [crates.io] page.
This command also has been requested by the community for a long time. You can find more discussion about this in [cargo#948].

Another motivation is to make the workflow of finding and evaluating crates more efficient. In the current workflow, users can search for crates using `cargo search`, but then they have to go to the web to get more information about the crate. This is not very efficient, especially if the user is just trying to get a quick overview of the crate. This would allow users to quickly get information about a crate without having to leave the terminal.

[crates.io]: https://crates.io
[cargo#948]: https://github.com/rust-lang/cargo/issues/948

Example usage:

```console
./target/debug/cargo info clap --verbose
  Credential cargo:token get crates-io
clap #argument #cli #arg #parser #parse
A simple to use, efficient, and full-featured Command Line Argument Parser
version: 4.5.8 (latest 4.5.9)
license: MIT OR Apache-2.0
rust-version: 1.74
documentation: https://docs.rs/clap/4.5.8
repository: https://github.com/clap-rs/clap
crates.io: https://crates.io/crates/clap/4.5.8
features:
 +default         = [std, color, help, usage, error-context, suggestions]
  color           = [clap_builder/color]
  error-context   = [clap_builder/error-context]
  help            = [clap_builder/help]
  std             = [clap_builder/std]
  suggestions     = [clap_builder/suggestions]
  usage           = [clap_builder/usage]
  cargo           = [clap_builder/cargo]
  debug           = [clap_builder/debug, clap_derive?/debug]
  deprecated      = [clap_builder/deprecated, clap_derive?/deprecated]
  derive          = [dep:clap_derive]
  env             = [clap_builder/env]
  string          = [clap_builder/string]
  unicode         = [clap_builder/unicode]
  unstable-doc    = [clap_builder/unstable-doc, derive]
  unstable-styles = [clap_builder/unstable-styles]
  unstable-v5     = [clap_builder/unstable-v5, clap_derive?/unstable-v5, deprecated]
  wrap_help       = [clap_builder/wrap_help]
dependencies:
 +clap_builder@=4.5.8
  clap_derive@=4.5.8
owners:
  kbknapp (Kevin K.)
  github:rust-cli:maintainers (Maintainers)
  github:clap-rs:admins (Admins)
note: to see how you depend on clap, run `cargo tree --invert --package clap@4.5.8`
```

<img width="1425" alt="image" src="https://github.com/user-attachments/assets/e0813c45-624f-417c-a61d-eda03f9ab5ed">

*note:* this is showing the `--verbose` output to show every thing the user can possibly see.  Normal operation does not include
- dependencies

## Detailed design

| Content                                                                    | Explanation                         | Why                                                                               |
|----------------------------------------------------------------------------|-------------------------------------|-----------------------------------------------------------------------------------|
| clap                                                                       | Name                                | The basic information.                                                            |
| #argument #cli #arg #parser #parse                                         | Keywords (clickable)                           | It's more like a category, which you can use to search for relevant alternatives. |
| A simple to use, efficient, and full-featured Command Line Argument Parser | Description                         | The basic information.                                                            |
| version: 4.5.8 (latest 4.5.9)                                                           | Version                             | The basic information.                                                            |
| license: MIT OR Apache-2.0                                                 | License                             | When choosing a crate, it is crucial to consider the license.                     |
| rust-version: 1.74                                                       | MSRV                                | When choosing a crate, it is crucial to make sure it can work with your MSRV.     |
| documentation: <https://docs.rs/clap/4.5.8>                               | Documentation Link                  | Use these links can find more docs and information.                               |
| repository: <https://github.com/clap-rs/clap>                              | Repo Link                           | Use these links can find more docs and information.                               |
| crates.io: https://crates.io/crates/clap/4.5.8                             | crates.io Link                           | Use these links can find more docs and information.                               |
| features:                                                                  | Default Features And Other Features | It helps for enabling features.                                                   |
| dependencies:                                                              | All dependencies                    | It indicates what it depends on.                                                  |
| owners:                                                                    | Owners                              | It indicates who maintains the crate.                                             |
| note: to see how you depend on clap, run `cargo tree --invert --package clap@4.5.8`                                                                   | A note for cargo tree command                              | It will prompt the user that the package is depended on under the workspace, and the dependencies can be viewed using the cargo tree command.           |

## Rendering features

- For features enabled by users, a + prefix and colored output are now used for better visibility.
- For features enabled automatically, colored output is used to distinguish them.
- For disabled features, non-colored output is used to clearly indicate their status.

## Rendering deps

Only show dependencies in verbose mode.

- For dependencies required by the package, a + prefix and colored output are now used for better visibility.
- For dependencies that are optional and activated, colored output is used to distinguish them.
- For dependencies that are optional and not activated, non-colored output is used to clearly indicate their status.

# Some important notes

## Downloading the crate from any Cargo compatible registry

The `cargo info` command will download the crate from any Cargo compatible registry. It will then extract the information from the `Cargo.toml` file and display it in the terminal.

If the crate is already in the local cache, it will not download the crate again. It will get the information from the local cache.

## Pick the correct version from the workspace

When executed in a workspace directory, the cargo info command chooses the version that the workspace is currently using.

If there's a lock file available, the version from this file will be used. In the absence of a lock file, the command attempts to select a version that is compatible with the Minimum Supported Rust Version (MSRV). And the lock file will be generated automatically.

The following hierarchy is used to determine the MSRV:

- First, the MSRV of the parent directory package is checked, if it exists.
- If the parent directory package does not specify an MSRV, the minimal MSRV of the workspace is checked.
- If neither the workspace nor the parent directory package specify an MSRV, the version of the current Rust compiler (rustc --version) is used.

# Prior art

## NPM

[npm] has a similar command called `npm info`.
For example:

```console
$ npm info lodash

lodash@4.17.21 | MIT | deps: none | versions: 114
Lodash modular utilities.
https://lodash.com/

keywords: modules, stdlib, util

dist
.tarball: https://registry.npmjs.org/lodash/-/lodash-4.17.21.tgz
.shasum: 679591c564c3bffaae8454cf0b3df370c3d6911c
.integrity: sha512-v2kDEe57lecTulaDIuNTPy3Ry4gLGJ6Z1O3vE1krgXZNrsQ+LFTGHVxVjcXPs17LhbZVGedAJv8XZ1tvj5FvSg==
.unpackedSize: 1.4 MB

maintainers:
- mathias <mathias@qiwi.be>
- jdalton <john.david.dalton@gmail.com>
- bnjmnt4n <benjamin@dev.ofcr.se>

dist-tags:
latest: 4.17.21

published over a year ago by bnjmnt4n <benjamin@dev.ofcr.se>
```

[npm]: https://www.npmjs.com/

## Poetry

[Poetry] has a similar command called `poetry show`.

For example:

```console
$ poetry show pendulum

name        : pendulum
version     : 1.4.2
description : Python datetimes made easy

dependencies
 - python-dateutil >=2.6.1
 - tzlocal >=1.4
 - pytzdata >=2017.2.2

required by
 - calendar >=1.4.0
```

[Poetry]: https://python-poetry.org/

# insta-stable

As `@weihanglo` mentioned in https://github.com/rust-lang/cargo/pull/14141#issuecomment-2189223552, commands that shadow third-party commands tend to be insta-stabilized to avoid an intermediate period where users can't access the third-party command (built-ins get priority) nor the built-in command (requires nightly)

For the cargo-info command, there are two commands that this would shadow
- [cargo-information](https://github.com/hi-rustin/cargo-information): hasn't been around too long and only has 4k downloads
- [cargo-info](https://gitlab.com/imp/cargo-info) : been around longer and has 63k downloads

We might be able to get away with having this unstable but starting from the assumption of insta-stabilization.

weihanglo committed:
test: show `trim-paths=diagnostics` doesnt works for rustdoc

It should be working with the next commit.

bors committed:
Auto merge of #14380 - jemtucker:fix-14379, r=epage

Use longhand gitoxide path-spec patterns

Closes #14379

bors committed:
Auto merge of #14389 - weihanglo:rustdoc, r=epage

feat(trim-paths): rustdoc supports trim-paths for diagnostics

weihanglo committed:
chore: downgrade to openssl v1.1.1 (again)

weihanglo committed:
chore: pin also openssl crate to 0.10.57

bors committed:
Auto merge of #14391 - weihanglo:downgrade-openssl, r=epage

chore: downgrade to openssl v1.1.1 (again)

It happened again in <https://github.com/rust-lang/cargo/pull/14299>.
See <https://github.com/rust-lang/cargo/issues/13546#issuecomment-2285181505> and <https://github.com/rust-lang/cargo/pull/13731>.

shannmu committed:
feat: Add custom completer for completing example names

ifropc committed:
lockfile path tests refactoring

ifropc committed:
doc: convert comments to rustdoc in workspace

ifropc committed:
feat: lockfile path move create logic into write_pkg_lockfile

epage committed:
refactor(update): Consolidate latest message rendering

epage committed:
refactor(update): Consolidate latest calculation

epage committed:
refactor(update): Extract rust-version lookup

akonradi committed:
Fix MSRV for workspace .package and .dependencies

The feature was stabilized in 1.64.0.

bors committed:
Auto merge of #14400 - akonradi:patch-1, r=epage

Fix MSRV for workspace .package and .dependencies

The feature was stabilized in 1.64.0

### What does this PR try to resolve?

Fixes a bug in the documentation.

### How should we test and review this PR?

The [release notes for 1.64.0](https://github.com/rust-lang/rust/releases/tag/1.64.0) list the features as stabilized.

### Additional information

This fixes a mismatch between this documentation and the note at the top of [the RFC](https://github.com/rust-lang/rfcs/blob/46781d0b1de3a07b10c2066d371ba7de2e209f15/text/2906-cargo-workspace-deduplicate.md).

epage committed:
fix(update): Only show 'latest' messages for registry sources

epage committed:
feat(update): Report when incompatible-rust-version packages are selected

In discussin this in #13873, it highlighted that we need to make sure we
tell people when we get in this state.

I decided to keep "latest" and "required rust version" messages mutually
exclusive to avoid too much noise.  I gave "required rust version"
higher precedence as its the more critical to operation and, if you are
using an MSRV-incompatible package, it likely is "latest" already.

I was tempted to change colors to make "required rust version" stand out
from "latest" but was unsure what direction to go, so I held off.
Options included
- red for "required rust version", yellow for "latest"
- yellow for "required rust version", nothing for "latest"

There is also more discussion on how to format "latest" at #13908.

bors committed:
Auto merge of #14397 - Ifropc:doc-clean, r=epage

doc: convert comments to rustdoc in workspace

Converting workspace comments into docs and adding docs on members that were missing it.
[Original discussion](https://github.com/rust-lang/cargo/pull/14326#discussion_r1702452550)

bors committed:
Auto merge of #14360 - dpaoliello:basepath1, r=epage

Implement base paths (RFC 3529) 1/n: path dep and patch support

RFC: https://github.com/rust-lang/rfcs/pull/3529
Tracking Issue: https://github.com/rust-lang/cargo/issues/14355

This PR add support in path dependencies and `patch` table and the `workspace` built-in path base.

epage committed:
test: migrate metabuild json to snapbox

epage committed:
test: migrate doc json to snapbox

epage committed:
chore: Update snapbox

Commit:
Add misbehaving test for packaging unpublishable workspace

bors committed:
Auto merge of #14402 - epage:snap, r=weihanglo

test: Migrate some json tests to snapbox

### What does this PR try to resolve?

This is part of #14039

### How should we test and review this PR?

### Additional information

This was unblocked because of assert-rs/snapbox#350

epage committed:
test: Migrate old_cargos to snapbox

This is part of #14039

linyihai committed:
test: add `opt_version_req_mathes_prerelease` test

linyihai committed:
fix: Limiting prerelease match semantics to use only on `OptVersionReq::Req`

ifropc committed:
test: lockfile path test refactoring

epage commented on src/cargo/util/command_prelude.rs:
Again, we should be using Cargo's APIs to look up the actual example names, rather than trying to guess

bors committed:
Auto merge of #14326 - Ifropc:5707-lock-path, r=weihanglo

Add `--lockfile-path` flag

This change implements a new `--lockfile-path` proposed in #5707 .

Functionality added:
- Add `--lockfile-path <PATH>` to all commands that support `manifest-path` with exception of:
   - `locate-project` (doesn't use lock file)
   - `verify-project` (is deprecated)
   - `read-manifest` (doesn't use lock file)
- Behind -Zunstable-options and docs
   - The flag's docs / `--help` has (unstable) in them
- `<PATH>` must end with `Cargo.lock`. If specified path doesn't exist (or parent director(ies), create all the parent directories and the lockfile itself

Implementation TLDR: add `requested_lockfile_path` into `Workspace` and set it on `workspace(gctx)` call (setting from the context)
Update `lockfile.lock_root()` to respect `requested_lockfile_path` (if set)
Add test cases covering all affected commands. Tested creating lockfile, reading lockfile, overriding default (`./Cargo.lock`) lockfile, symlink tests. Extra tests for package to make sure pinned versions from path's lockfile are respected (i.e. double check correct lockfile is used)
I doubt this flag will be used for any command that's not read-only, but I tried to cover all the commands.

ctz committed:
Correct diagnostic for `TomlDebugInfo`

This is missing the friendly aliases introduced in 3dbb474c.

bors committed:
Auto merge of #14413 - ctz:master, r=epage

Correct diagnostic for `TomlDebugInfo`

This is missing the friendly aliases introduced in 3dbb474c.

### What does this PR try to resolve?

I wrote:

```
[profile.bench]
debug = "all"
```

I got this diagnostic:

```
error: invalid value: string "all", expected a boolean, 0, 1, 2, "line-tables-only", or "line-directives-only"
  --> Cargo.toml:31:9
   |
31 | debug = "all"
   |         ^^^^^
   |
```

I meant to write "full", as [documented here](https://doc.rust-lang.org/cargo/reference/profiles.html#debug), but that wasn't mentioned in the diagnostic.

### How should we test and review this PR?

Updated the existing tests for this diagnostic.

bors committed:
Auto merge of #14410 - epage:tests, r=weihanglo

test: Migrate old_cargos to snapbox

This is part of #14039

bors committed:
Auto merge of #14401 - epage:required-rust, r=weihanglo

feat(update): Report when incompatible-rust-version packages are selected

### What does this PR try to resolve?

In discussin this in #13873, it highlighted that we need to make sure we
tell people when incompatible-rust-version packages are selected.

I decided to keep "latest" and "required rust version" messages mutually
exclusive to avoid too much noise.  I gave "required rust version"
higher precedence as its the more critical to operation and, if you are
using an MSRV-incompatible package, it likely is "latest" already.

### How should we test and review this PR?

### Additional information

I was tempted to change colors to make "required rust version" stand out
from "latest" but was unsure what direction to go, so I held off.
Options included
- red for "required rust version", yellow for "latest"
- yellow for "required rust version", nothing for "latest"

There is also more discussion on how to format "latest" at #13908.

0xPoe committed:
fix: remove list owners feature of info subcommand

ifropc committed:
test: lockfile path test refactoring

- Remove "execs" method

ifropc committed:
test: lockfile path unstable feature tests

bors committed:
Auto merge of #14417 - Ifropc:lockfile-path-tests, r=weihanglo

Lockfile path tests (follow-up)

Addressing comments by [Weihang](https://github.com/rust-lang/cargo/pull/14326#issuecomment-2293664459) and [Ed](https://github.com/rust-lang/cargo/pull/14326#issuecomment-2293671992) from #14326

bors committed:
Auto merge of #14418 - Rustin170506:rustin-patch-info-owners, r=weihanglo

fix: remove list owners feature of info subcommand

weihanglo committed:
chore: update label trigger for Command-info

ifropc committed:
doc: add lockfile-path unstable doc section

ifropc committed:
doc: update lockfile-path tracking issue

bors committed:
Auto merge of #14424 - Ifropc:lockfile-tracking-issue, r=weihanglo

doc: update lockfile-path tracking issue

Update tracking issue of lockfile path feature to https://github.com/rust-lang/cargo/issues/14421
(in documentation and code)
r? `@weihanglo`

ShashiSugrim committed:
Fix issue ' cargo t --doc does not respect --color when a compilation error occurs #14403 '

Commit:
Use the shared source building in package

This changes the registry validation slightly, adding in a check
forbidding implicit source replacement. This affects the tests (which
configure a dummy registry for source replacement), so we also weaken
the checks by only erroring for registry issues when there are actually
local dependencies.

Commit:
Allow unpublishable crates to be packaged

Commit:
Split infer_registry into inference and validation parts.

Commit:
Move infer_registry into a more public spot

bors committed:
Auto merge of #14423 - Ifropc:lockfile-doc, r=weihanglo

doc: add lockfile-path unstable doc section

Adds lockfile-path into the unstable doc section as part of tracking issue https://github.com/rust-lang/cargo/issues/14421 tasks

bors committed:
Auto merge of #14422 - weihanglo:label, r=epage

chore: update label trigger for Command-info

dpaoliello committed:
[mdman] Add a test for a code block split across multiple lines in an options template

dpaoliello committed:
[mdman] Normalize newlines when rendering options

0xPoe committed:
test: add a regression test for Issue 14409

Signed-off-by: hi-rustin <29879298+hi-rustin@users.noreply.github.com>

bors committed:
Auto merge of #14430 - Rustin170506:rustin-patch-14409, r=epage

test: add a regression test for Issue 14409

bors committed:
Auto merge of #14412 - linyihai:fix-pre-release, r=epage

fix: Limiting pre-release match semantics to use only on `OptVersionReq::Req`

### What does this PR try to resolve?

The prerelease matches semantics should be only used on `OptVersionReq::Req`, but it dosn't.  The other `OptVersionReq` types have specify matches logic already,  for example a `OptVersionReq::Precise` will failed on `matches_prerelease`, see https://github.com/rust-lang/cargo/pull/14140#discussion_r1682777074

### How should we test and review this PR?

the first commit added the test, the second commit updated the test and fixed the issue.

### Additional information

bors committed:
Auto merge of #14408 - tweag:infer-registry-ignore-unpublishable, r=epage

Be more permissive while packaging unpublishable crates.

This PR allows for packaging workspaces that include `publish = false` crates, in some circumstances:

- unpublishable crates are ignored when inferring the publish registry
- when checking whether the publish registry is valid, we ignore unpublishable crates
- we don't put unpublishable crates in the local registry overlay, so if any workspace crates depend on an unpublishable crate then they will fail to verify.

This PR also contains a refactor, moving the registry inference logic to `registry/mod.rs`, where it will be reused by the upcoming publish-workspace feature. I put the refactor and the logic changes in different commits.

Fixes #14356

bors committed:
Auto merge of #14425 - ShashiSugrim:shashi/issue14403, r=epage

fix: doctest respects Cargo's color options

### What does this PR try to resolve?

Explain the motivation behind this change.

This commit fixes the issue where cargo's test command doesn't respect the color parameter when it gets passed in like this command: `cargo t --color never --doc -- --color never`

Fixes #14403
### How should we test and review this PR?
Test on a basic rust project, with a file called lib.rs that looks like this
```
/// ```
/// bar
/// ```
pub fn foo() {}
#[cfg(test)]
mod tests {
    #[test]
    fn foo() {
        bar
    }
}
```
You can try to replicate the same commands `@zacknewman` used in the description of this issue.
`cargo t --color never --doc -- --color never`

You will see that compared to the official build of cargo, this build will respect the --color argument

bors committed:
Auto merge of #14428 - dpaoliello:mdman, r=epage

[mdman] Normalize newlines when rendering options

When using `cargo build-man` on Windows, I noticed that some files I didn't modify were getting modified, specifically they had an extra space added in the middle of a code block in an option description.

The root cause appears to be that if a code block in a handlebars template was split across multiple lines, then both the `\r` and the `\n` were being converted to spaces.

The fix for this is to normalize newlines to `\n` after templates are expanded by handlebars but before the individual formatters render the block as an option.

epage committed:
refactor(update): Move the allocation out of PackageDiff

epage committed:
refactor(update): Prefer Copy over references

epage committed:
refactor(update): Centralize "what is a change"

This is prep for consolidating a lot of code *and* tracking more
information so we can have a richer output.

epage committed:
fix(update): Show required rust-version in one more case

epage committed:
refactor(update): Be clearer, consistent in variable name

epage committed:
refactor(update): Remove unnecesary derefs

epage committed:
refactor(update): Be consistent in what source we interact with

epage committed:
fix(update): Add missing required rust-version note for git

epage committed:
refactor(update): Centralize status key

epage committed:
refactor(update): Centralize status style

weihanglo committed:
test: show runtime-dep and build-dep collides

See rust-lang/cargo 14253

epage committed:
refactor(update): Consolidate change rendering

weihanglo committed:
fix: `-Cmetadata` includes whether extra rustflags is same as host

`-Ztarget-applies-to-host --config target-applies-to-host=false` make it
possible to build two units for the same output directory with different
rustflags (one with artifact/target rustflags, one with none/host rust
flags). Those flags aren't included in the `-Cmetadata` hash which is
used to disambiguate different versions of crates, resulting in a conflict.
The actual error being produced appears to be the result of two invocations
of `rustc` racing and clobbering each other.

While we don't hash RUSTFLAGS because it may contain absolute paths that
hurts reproducibility, we track whether a unit's RUSTFLAGS is from host
config, so that we can generate a different metadata hash for runtime
and compile-time units.

linyihai committed:
test: add more test for pre-release matches semantic

kornelski committed:
More helpful missing feature error message

epage committed:
refactor(update): Consolidate status messages

This builds on the prior work to consolidate everything, simplifying the
code and making it clearer what behavior differences exist between
change kinds.

epage committed:
fix(resolve): Dont show locking workspace members

This is for `cargo generate-lockfile` and when syncing the lockfile with
the manifest.
We still show it for `cargo update` because of `cargo update
--workspace`.

We hacked around this previously by filtering out the `num_pkgs==1` case
for single packages but this didn't help with workspaces.

bors committed:
Auto merge of #14432 - weihanglo:target-applies-to-host, r=epage

fix: -Cmetadata includes whether extra rustflags is same as host

### What does this PR try to resolve?

Fixes #14253

The root cause is described in <https://github.com/rust-lang/cargo/issues/14253#issuecomment-2299595480>:

> `-Ztarget-applies-to-host --config target-applies-to-host=false` make it possible to build two units for the same output directory with different rustflags (one with artifact/target rustflags, one with none/host rust flags). Those flags aren't included in the `-Cmetadata` hash which is used to disambiguate different versions of crates, resulting in a conflict. The actual error being produced appears to be the result of two invocations of `rustc` racing and clobbering each other…

While we don't hash RUSTFLAGS because it may contain absolute paths that
hurts reproducibility, we track whether a unit's RUSTFLAGS is from host
config, so that we can generate a different metadata hash for runtime
and compile-time units.

### How should we test and review this PR?

This hack is only enabled when `target-applies-to-host=false` and when compile kind is the host platform, so the shouldn't affect any stable behavior.

kornelski committed:
Tests rely on absence of RUST_BACKTRACE

bors committed:
Auto merge of #14441 - kornelski:env-leak, r=epage

Tests rely on absence of RUST_BACKTRACE

I have it set in my default environment, and it causes tests to fail

bors committed:
Auto merge of #14440 - epage:report, r=Muscraft

refactor(update): Prepare for smarter update messages

### What does this PR try to resolve?

This is to make it easier to
- Make #14401 path aware
- Work on #13908, depending on what is decided
- Improve the heuristics for when we show `Locking` messages

There are fixes along the way that were found by making the code more consistent in prep for consolidating it, mostly for #14401.

### How should we test and review this PR?

Along the way, some odd code structures are used for the sake of making the diff easier to follow.  These get cleaned up towards the end

I didn't add tests for the fixes because of the code consolidation that happens that causes us to cover more real-world scenarios with fewer tests.

### Additional information

bors committed:
Auto merge of #14305 - linyihai:matches-prerelease-semantic, r=epage

feat: Add matches_prerelease semantic

### What does this PR try to resolve?
One implementaion for https://github.com/rust-lang/cargo/issues/13290

Thanks to `@Eh2406` feedback, a working version came out, and I think it should be able to test under the nightly feature.

This PR proposes a `matches_prerelease semantic`.

| req              | matches             | matches_prerelease     | matches_prerelease_mirror_node [<sub>2<sub>](#mirror-node) |
|------------------|---------------------|------------------------| ----------------------------------|
| `Op::Exact`      |                     |                        |                                   |
| =I.J.K           | =I.J.K              | >=I.J.K, <I.J.(K+1)-0  | >=I.J.K, <I.J.(K+1)-0             |
| =I.J             | >=I.J.0, <I.(J+1).0 | >=I.J.0, <I.(J+1).0-0  | >=I.J.0-0, <I.(J+1).0-0           |
| =I               | >=I.0.0, <(I+1).0.0 | >=I.0.0, <(I+1).0.0-0  | >=I.0.0-0, <(I+1).0.0-0           |
| `Op::Greater`    |                     |                        |                                   |
| >I.J.K           | >I.J.K              | >I.J.K                 | >I.J.K                            |
| >I.J             | >=I.(J+1).0         | >=I.(J+1).0-0          | >=I.(J+1).0-0                     |
| >I               | >=(I+1).0.0         | >=(I+1).0.0-0          | >=(I+1).0.0-0                     |
| `Op::GreaterEq`  |                     |                        |                                   |
| >=I.J.K          | >=I.J.K             | >=I.J.K                | >=I.J.K                           |
| >=I.J            | >=I.J.0             | >=I.J.0                | >=I.J.0-0                         |
| >=I              | >=I.0.0             | >=I.0.0                | >=I.0.0-0                         |
| `Op::Less`       |                     |                        |                                   |
| <I.J.K           | <I.J.K              | <I.J.K or I.J.K-0 depends [<sub>1<sub>](#op-less) | <I.J.K |
| <I.J             | <I.J.0              | <I.J.0-0               | <I.J.0-0                          |
| <I               | <I.0.0              | <I.0.0-0               | <I.0.0-0                          |
| `Op::LessEq`     |                     |                        |                                   |
| <=I.J.K          | <=I.J.K             | <=I.J.K                | <=I.J.K                           |
| <=I.J            | <I.(J+1).0          | <I.(J+1).0-0           | <I.(J+1).0-0                      |
| <=I              | <(I+1).0.0          | <(I+1).0.0-0           | <(I+1).0.0-0                      |
| `Op::Tilde`      |                     |                        |                                   |
| ~I.J.K           | >=I.J.K, <I.(J+1).0 | >=I.J.K, <I.(J+1).0-0  | >=I.J.K, <I.(J+1).0-0             |
| ~I.J             | =I.J                | >=I.J.0, <I.(J+1).0-0  | >=I.J.0, <I.(J+1).0-0             |
| ~I               | =I                  | >=I.0.0, <(I+1).0.0-0  | >=I.0.0, <(I+1).0.0-0             |
| `Op::Caret`      |                     |                        |                                   |
| ^I.J.K (for I>0) | >=I.J.K, <(I+1).0.0 | >=I.J.K, <(I+1).0.0-0  | >=I.J.K, <(I+1).0.0-0             |
| ^0.J.K (for J>0) | >=0.J.K, <0.(J+1).0 | >=0.J.K,  <0.(J+1).0-0 | >=0.J.K-0, <0.(J+1).0-0           |
| ^0.0.K           | =0.0.K              | >=0.0.K,  <0.0.(K+1)-0 | >=0.J.K-0, <0.(J+1).0-0           |
| ^I.J             | ^I.J.0              | >=I.J.0,  <(I+1).0.0-0 | >=I.J.0-0, <(I+1).0.0-0           |
| ^0.0             | =0.0                | >=0.0.0, <0.1.0-0      | >=0.0.0-0, <0.1.0-0               |
| ^I               | =I                  | >=I.0.0, <(I+1).0.0-0  | >=I.0.0-0, <(I+1).0.0-0           |
| `Op::Wildcard`   |                     |                        |                                   |
| `I.J.*`          | =I.J                | >=I.J.0, <I.(J+1).0-0  | >=I.J.0-0, <I.(J+1).0-0           |
| `I.*` or `I.*.*` | =I                  | >=I.0.0, <(I+1).0.0-0  | >=I.0.0-0, <(I+1).0.0-0           |

Notes:

<div id="op-less"></div>

- `<I.J.K`: This is equivalent to `<I.J.K-0` if no lower bound or the lower bound isn't pre-release, otherwise this is equivalent to `<I.J.K`.

<div id="mirror-node"></div>

- [matches_prerelease_mirror_node](https://github.com/dtolnay/semver/pull/321/commits/3464fd136a180fd64c6b0a394ccb0479ccb8d967) is yet another  implementation of [node semver compatibility](https://github.com/npm/node-semver?tab=readme-ov-file#semver1----the-semantic-versioner-for-npm) (with [includePrerelease=true](https://github.com/npm/node-semver?tab=readme-ov-file#functions)) test. This is extrapolated from the test cases and is not necessarily the same as the node implementation

Besides, the proposed semtantic left a [unsolved issuse ](https://github.com/dtolnay/semver/pull/321#issuecomment-2251058953), I don't have clear idea to handle it.

### How should we test and review this PR?
The tests in `src/cargo/util/semver_eval_ext.rs` are designed to reflect current semantic.

### Additional information
Migrated from https://github.com/dtolnay/semver/pull/321

TBO, this feature was not conceived in advance plus the semantic is unclear at first.  I need to experiment step by step and find out what I think makes sense.  My experiment can be seen this [comment](https://github.com/rust-lang/cargo/issues/13290#issuecomment-2250051428).

epage committed:
refactor(resolve): Allow lookups on changes by id

epage committed:
fix(resolve): Report incompatible packages with precise Rust version

The nesting in `annotate_required_rust_version` might seem weird but its
for a follow up commit.

epage committed:
test(resolve): Explicitly verify msrv report

epage committed:
feat(resolve): Report incompatible with rustc when MSRV-resolve disabled

shannmu committed:
test(auto_complete): Add codebase of auto-complete

bors committed:
Auto merge of #14436 - kornelski:where-are-the-features, r=epage

More helpful missing feature error message

The "none of the selected packages contains these features" error does not say which packages are selected, and does not say which packages do have the requested features.

If any package in the workspace has the requested feature (matching exactly), the error suggests selecting that package, instead of trying alternative spellings.

shannmu committed:
test: Add test case for completing example names

Mark-Simulacrum committed:
Log details of failure if no errors were seen

bors committed:
Auto merge of #14453 - Mark-Simulacrum:debug-no-errors, r=weihanglo

Log details of failure if no errors were seen

### What does this PR try to resolve?

Crater has started (some time ago, though exact point is not known) to see failures in crates without logging an error. We're suspecting that there's something wrong in Cargo or rustc, but it's also possible that Crater itself is at fault. In any case, this should provide more details.

cc https://github.com/rust-lang/crater/issues/733

### How should we test and review this PR?

My hope is that this is reasonable to land without additional tests -- as-is, we expect it to never happen (there's even a debug assert later in the same code path) but it seems it *is* happening sometimes in the wild.

epage committed:
fix(resolve): Elevate the MSRV severity style now that its precise

bors committed:
Auto merge of #14445 - epage:locked, r=weihanglo

fix(resolve): Dont show locking workspace members

### What does this PR try to resolve?
This is for `cargo generate-lockfile` and when syncing the lockfile with
the manifest.
We still show it for `cargo update` because of `cargo update
--workspace`.

We hacked around this previously by filtering out the `num_pkgs==1` case
for single packages but this didn't help with workspaces.

### How should we test and review this PR?

### Additional information

This builds on #14440

bors committed:
Auto merge of #14457 - epage:msrv, r=ehuss

fix(resolve): Report incompatible packages with precise Rust version

### What does this PR try to resolve?

In #14401, we reported about MSRV issues as if we were the resolver.  We can be smarter than that though and report precise MSRV information.  This allows us to elevate the message from color from yellow to red.

This is also prep work for telling users when a newer, MSRV-compatible version of a package is available.

### How should we test and review this PR?

The report function I added is a little odd because a follow up commit will update it to report when a package is incompatible with rustc when the MSRV resolver is disabled and do this on stable.

### Additional information

This builds on #14445 to improve #14401

bors committed:
Auto merge of #14459 - epage:msrv-rustc, r=Muscraft

feat(resolve): Report incompatible-with-rustc when MSRV-resolver is disabled

### What does this PR try to resolve?

This builds on #14457 to report when deps are incompatible with rustc when an MSRV-aware resolver is not used.
This affects stable code.

### How should we test and review this PR?

### Additional information

shannmu committed:
feat: Add native comlpetion with `CompleteEnv` and under the nightly features
